internal/controllers/core/portforward/reconciler.go

```diff
@@ -0,0 +1,199 @@
+package portforward
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+
+	"github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
+
+	"github.com/tilt-dev/tilt/internal/store/k8sconv"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/tilt-dev/tilt/internal/k8s"
+	"github.com/tilt-dev/tilt/internal/store"
+	"github.com/tilt-dev/tilt/pkg/logger"
+	"github.com/tilt-dev/tilt/pkg/model"
+)
+
+type Reconciler struct {
+	// until PortForwarding behavior is moved off the PortForwardSubscriber and onto
+	// the reconciler, we set the reconciler to active/not with a flag (there should
+	// be no PortForwards on the EngineState anyway, but this is a failsafe).
+	isActive bool
+
+	kClient k8s.Client
+
+	// map of PortForward object name --> running forward(s)
+	activeForwards map[string]portForwardEntry
+}
+
+func NewReconciler(kClient k8s.Client) *Reconciler {
+	return &Reconciler{
+		kClient:        kClient,
+		activeForwards: make(map[string]portForwardEntry),
+	}
+}
+
+func (r *Reconciler) SetActive() {
+	// Should only be used for tests; should be deprecated when migration is complete
+	r.isActive = true
+}
+
+// Figure out the diff between what's in the data store and
+// what port-forwarding is currently active.
+func (r *Reconciler) diff(ctx context.Context, st store.RStore) (toStart []portForwardEntry, toShutdown []portForwardEntry) {
+	state := st.RLockState()
+	defer st.RUnlockState()
+
+	statePFs := state.PortForwards
+
+	for name, existing := range r.activeForwards {
+		if _, onState := statePFs[name]; !onState {
+			// This port forward is no longer on the state, shut it down.
+			toShutdown = r.addToShutdown(toShutdown, existing)
+			continue
+		}
+	}
+
+	for name, desired := range statePFs {
+		existing, isActive := r.activeForwards[name]
+		if isActive {
+			// We're already running this PortForward -- do we need to do anything further?
+			if equality.Semantic.DeepEqual(existing.Spec, desired.Spec) && equality.Semantic.DeepEqual(existing.ObjectMeta, desired.ObjectMeta) {
```

Comment:
comparing the ObjectMeta like this isn't right...it's going to compare things you don't expect, like ResourceVersion

Reply:
Hmm fair. so the one thing i need to make sure stays the same is the ManifestName annotation -- would it be too weird to compare that directly? or should i duplicate that data on the state? or?